sdk/include/platform/sunburst/platform-usbdev.hh

```diff
@@ -0,0 +1,633 @@
+	 * or releasing the buffer.
+	 *
+	 * @param endpointId An out-parameter, to which the ID of the endpoint for
+	 * a recently-collected IN data packet will be written.
+	 * @param bufferId An out-parameter, to which the ID of the buffer for a
+	 * recently-collected IN data packet will be written.
+	 *
+	 * @returns 0 if successful, and non-zero otherwise.
+	 */
+	[[nodiscard]] int retrieve_collected_packet(uint8_t &endpointId,
+	                                            uint8_t &bufferId) volatile
+	{
+		constexpr uint32_t BufferIdMask =
+		  static_cast<uint32_t>(ConfigInField::BufferId);
+		uint32_t sent = inSent;
+
+		// Clear the first encountered packet sent indication.
+		for (endpointId = 0; endpointId < MaxEndpoints; endpointId++)
+		{
+			const uint32_t EndpointBit = 1u << endpointId;
+			if (sent & EndpointBit)
+			{
+				// Clear the `in_sent` bit for this specific endpoint, and
+				// indicate which buffer has been released.
+				inSent   = EndpointBit;
+				bufferId = (configIn[endpointId] & BufferIdMask);
+				return 0;
+			}
+		}
+
+		// If no packet sent indications were found, then fail.
+		return -1;
+	}
+
+	/**
+	 * Present a packet on the specified IN endpoint for collection by the USB
+	 * host controller.
+	 *
+	 * @param bufferId The buffer to use to store the packet.
+	 * @param endpointId The IN endpoint used to send the packet.
+	 * @param data The packet to be transmitted.
+	 * @param size The size of the packet.
+	 *
+	 * @returns 0 if successful, or non-zero if not successful.
+	 */
+	[[nodiscard]] int packet_send(uint8_t         bufferId,
+	                              uint8_t         endpointId,
+	                              const uint32_t *data,
+	                              uint8_t         size) volatile
+	{
+		// Transmission of zero length packets is common over USB
+		if (size > 0)
+		{
+			usbdev_transfer(buffers(bufferId), data, size, true);
+		}
+
+		constexpr uint32_t ReadyBit =
+		  static_cast<uint32_t>(ConfigInField::Ready);
+		configIn[endpointId] = bufferId | (size << 8);
+		configIn[endpointId] = configIn[endpointId] | ReadyBit;
+		return 0;
+	}
+
+	/**
+	 * Test for and collect the next received packet.
+	 *
+	 * @param bufferId An out-parameter storing the buffer ID used
+	 * @param endpointId An out-parameter storing the endpoint received on.
+	 * @param size An out-parameter storing the size of the received packet.
+	 * @param isSetup A boolean out-parameter. True means the received packet
+	 * was a SETUP packet, false means it was not.
+	 * @param destination A destination buffer to read the packet's data into.
+	 *
+	 * @returns 0 if successful, or non-zero if not successful.
+	 */
+	[[nodiscard]] int packet_receive(uint8_t  &bufferId,
+	                                 uint8_t  &endpointId,
+	                                 uint16_t &size,
+	                                 bool     &isSetup,
+	                                 uint32_t *destination) volatile
```

Comment:
This would be great to have, but we want to empty pop receive buffer as soon as possible. We can leave the data in the buffer to be read at a later time by someone with the correct endpoint permission, but we need to ensure it is made available again once it has been read. We'd need some way to track this. We also really don't want to be starving the availableSetupFifo, which could happen if an endpoint is allowed to neglect reading it's buffers.

I think this functionality would make sense in an intermediate library (or compartment if necessary).

Reply:
The functionality would need to be in another compartment, but I’m concerned that this API doesn’t give us what we need to implement that compartment.

Reply:
I think you're right. I've split the function into two more granular ones: https://github.com/CHERIoT-Platform/cheriot-rtos/pull/337/commits/91e376b12038531fcd9695fb8ce9c9140ebc35c8 .

Does this look good to you?

Reply:
Yup, that's nice.  Now a loop can sit and receive packets and pass them to the thread that is waiting for that event easily.

For some extra error checking, it would be nice if:

 - `ReceiveBufferInfo` had a move constructor and a deleted copy constructor, where the move constructor marks the source as invalid.
 - `packet_data_get ` took a `ReceiveBufferInfo&&`, so consumes the buffer.  


Not essential though.